WOS transform key not found error

```
Transforming for 30648 (row: 989)
2018-10-30T16:35:59-04:00 [DEBUG] 23464 : beginning Traject::Indexer#process with settings: {"writer_class_name"=>"Rialto::Etl::Writers::SparqlStatementWriter", "reader_class_name"=>"Rialto::Etl::Readers::NDJsonReader", "output_file"=>"pubs-data/wos-30648.sparql", "solr_writer.batch_size"=>100, "solr_writer.thread_pool"=>1, "processing_thread_pool"=>1, "log.batch_size.severity"=>"info", "allow_nil_values"=>false, "allow_duplicate_values"=>true, "allow_empty_fields"=>false}
2018-10-30T16:35:59-04:00 [ INFO] 23464 :    Traject::Indexer with 1 processing threads, reader: Rialto::Etl::Readers::NDJsonReader and writer: Rialto::Etl::Writers::SparqlStatementWriter
2018-10-30T16:36:04-04:00 [DEBUG] 23464 : Shutting down #processing mapper threadpool...
2018-10-30T16:36:05-04:00 [DEBUG] 23464 : #processing mapper threadpool shutdown complete.
2018-10-30T16:36:05-04:00 [ INFO] 23464 : finished Traject::Indexer#process: 5 records in 5.773 seconds; 0.9 records/second overall.
Extracting for 50908 (row: 990)
Transforming for 50908 (row: 990)
2018-10-30T16:36:07-04:00 [DEBUG] 23465 : beginning Traject::Indexer#process with settings: {"writer_class_name"=>"Rialto::Etl::Writers::SparqlStatementWriter", "reader_class_name"=>"Rialto::Etl::Readers::NDJsonReader", "output_file"=>"pubs-data/wos-50908.sparql", "solr_writer.batch_size"=>100, "solr_writer.thread_pool"=>1, "processing_thread_pool"=>1, "log.batch_size.severity"=>"info", "allow_nil_values"=>false, "allow_duplicate_values"=>true, "allow_empty_fields"=>false}
2018-10-30T16:36:07-04:00 [ INFO] 23465 :    Traject::Indexer with 1 processing threads, reader: Rialto::Etl::Readers::NDJsonReader and writer: Rialto::Etl::Writers::SparqlStatementWriter
2018-10-30T16:36:11-04:00 [ERROR] 23465 : Unexpected error on record <record #2 (pubs-data/wos-50908.ndj #2)>
    while executing (to_field "http://vivoweb.org/ontology/core#relatedBy" at lib/rialto/etl/configs/wos_to_sparql_statements.rb:116)

    Record: {"UID"=>"WOS:000398181700011", "static_data"=>{"summary"=>{"pub_info"=>{"coverdate"=>"APR 2017", "vol"=>30, "pubyear"=>2017, "sortdate"=>"2017-04-01", "has_abstract"=>"Y", "pubmonth"=>"APR", "pubtype"=>"Journal", "page"=>{"end"=>138, "begin"=>128, "page_count"=>11, "content"=>"128-138"}}, "names"=>{"count"=>2, "name"=>[{"seq_no"=>1, "orcid_id"=>"0000-0002-0956-8001", "role"=>"author", "full_name"=>"Gonzales, Patricia", "reprint"=>"Y", "addr_no"=>"1 2 4", "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "wos_standard"=>"Gonzales, P", "r_id"=>"O-3352-2017", "daisng_id"=>1389780, "first_name"=>"Patricia"}, {"seq_no"=>2, "orcid_id"=>"0000-0003-4421-3764", "role"=>"author", "full_name"=>"Ajami, Newsha K.", "addr_no"=>"2 3", "last_name"=>"Ajami", "display_name"=>"Ajami, Newsha K.", "wos_standard"=>"Ajami, NK", "r_id"=>"C-9151-2017", "daisng_id"=>2367950, "first_name"=>"Newsha K."}]}, "doctypes"=>{"doctype"=>"Article", "count"=>1}, "publishers"=>{"publisher"=>{"names"=>{"count"=>1, "name"=>{"seq_no"=>1, "role"=>"publisher", "full_name"=>"ELSEVIER SCIENCE BV", "addr_no"=>1, "display_name"=>"ELSEVIER SCIENCE BV"}}, "address_spec"=>{"city"=>"AMSTERDAM", "addr_no"=>1, "full_address"=>"PO BOX 211, 1000 AE AMSTERDAM, NETHERLANDS"}}}, "EWUID"=>{"WUID"=>{"coll_id"=>"WOS"}, "edition"=>[{"value"=>"WOS.SCI"}, {"value"=>"WOS.SSCI"}]}, "titles"=>{"count"=>6, "title"=>[{"type"=>"source", "content"=>"SUSTAINABLE CITIES AND SOCIETY"}, {"type"=>"source_abbrev", "content"=>"SUSTAIN CITIES SOC"}, {"type"=>"abbrev_iso", "content"=>"Sust. Cities Soc."}, {"type"=>"abbrev_11", "content"=>"SUS CIT SOC"}, {"type"=>"abbrev_29", "content"=>"SUSTAIN CITIES SOC"}, {"type"=>"item", "content"=>"An integrative regional resilience framework for the changing urban water paradigm"}]}}, "item"=>{"xsi:type"=>"itemType_wos", "coll_id"=>"WOS", "ids"=>{"avail"=>"N", "content"=>"EQ6HF"}, "xmlns:xsi"=>"http://www.w3.org/2001/XMLSchema-instance", "bib_pagecount"=>{"type"=>"Journal", "content"=>301}, "keywords_plus"=>{"count"=>10, "keyword"=>["LOS-ANGELES", "SUSTAINABILITY ASSESSMENT", "SOCIO-HYDROLOGY", "CLIMATE-CHANGE", "MANAGEMENT", "SYSTEMS", "CALIFORNIA", "DIVERSITY", "GOVERNANCE", "DROUGHT"]}, "bib_id"=>"30: 128-138 APR 2017"}, "fullrecord_metadata"=>{"addresses"=>{"count"=>4, "address_name"=>[{"names"=>{"count"=>1, "name"=>{"seq_no"=>1, "role"=>"author", "full_name"=>"Gonzales, Patricia", "reprint"=>"Y", "addr_no"=>1, "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "wos_standard"=>"Gonzales, P", "daisng_id"=>1389780, "first_name"=>"Patricia"}}, "address_spec"=>{"zip"=>{"location"=>"AP", "content"=>94305}, "country"=>"USA", "city"=>"Stanford", "street"=>"473 Via Ortega,Room 314", "addr_no"=>1, "organizations"=>{"organization"=>["Stanford Univ", {"pref"=>"Y", "content"=>"Stanford University"}], "count"=>2}, "full_address"=>"Stanford Univ, Dept Civil & Environm Engn, 473 Via Ortega,Room 314, Stanford, CA 94305 USA", "state"=>"CA", "suborganizations"=>{"count"=>1, "suborganization"=>"Dept Civil & Environm Engn"}}}, {"names"=>{"count"=>2, "name"=>[{"seq_no"=>1, "role"=>"author", "full_name"=>"Gonzales, Patricia", "reprint"=>"Y", "addr_no"=>2, "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "wos_standard"=>"Gonzales, P", "daisng_id"=>1389780, "first_name"=>"Patricia"}, {"seq_no"=>2, "role"=>"author", "full_name"=>"Ajami, Newsha K.", "addr_no"=>2, "last_name"=>"Ajami", "display_name"=>"Ajami, Newsha K.", "wos_standard"=>"Ajami, NK", "daisng_id"=>2367950, "first_name"=>"Newsha K."}]}, "address_spec"=>{"zip"=>{"location"=>"AP", "content"=>94305}, "country"=>"USA", "city"=>"Stanford", "street"=>"473 Via Ortega,Room 117", "addr_no"=>2, "organizations"=>{"organization"=>["Stanford Univ", {"pref"=>"Y", "content"=>"Stanford University"}], "count"=>2}, "full_address"=>"Stanford Univ, ReNUWIt Engn Res Ctr, 473 Via Ortega,Room 117, Stanford, CA 94305 USA", "state"=>"CA", "suborganizations"=>{"count"=>1, "suborganization"=>"ReNUWIt Engn Res Ctr"}}}, {"names"=>{"count"=>1, "name"=>{"seq_no"=>2, "role"=>"author", "full_name"=>"Ajami, Newsha K.", "addr_no"=>3, "last_name"=>"Ajami", "display_name"=>"Ajami, Newsha K.", "wos_standard"=>"Ajami, NK", "daisng_id"=>2367950, "first_name"=>"Newsha K."}}, "address_spec"=>{"zip"=>{"location"=>"AP", "content"=>94305}, "country"=>"USA", "city"=>"Stanford", "street"=>"473 Via Ortega,Room 204", "addr_no"=>3, "organizations"=>{"organization"=>["Stanford Univ", {"pref"=>"Y", "content"=>"Stanford University"}], "count"=>2}, "full_address"=>"Stanford Univ, Woods Inst Environm, 473 Via Ortega,Room 204, Stanford, CA 94305 USA", "state"=>"CA", "suborganizations"=>{"count"=>1, "suborganization"=>"Woods Inst Environm"}}}, {"names"=>{"count"=>1, "name"=>{"seq_no"=>1, "role"=>"author", "full_name"=>"Gonzales, Patricia", "reprint"=>"Y", "addr_no"=>4, "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "wos_standard"=>"Gonzales, P", "daisng_id"=>1389780, "first_name"=>"Patricia"}}, "address_spec"=>{"zip"=>{"location"=>"AP", "content"=>94305}, "country"=>"USA", "city"=>"Stanford", "street"=>"473 Via Ortega,Room 204", "addr_no"=>4, "full_address"=>"473 Via Ortega,Room 204, Stanford, CA 94305 USA", "state"=>"CA"}}]}, "category_info"=>{"subheadings"=>{"count"=>1, "subheading"=>"Technology"}, "subjects"=>{"subject"=>[{"ascatype"=>"traditional", "code"=>"FA", "content"=>"Construction & Building Technology"}, {"ascatype"=>"traditional", "code"=>"GS", "content"=>"GREEN & SUSTAINABLE SCIENCE & TECHNOLOGY"}, {"ascatype"=>"traditional", "code"=>"ID", "content"=>"Energy & Fuels"}, {"ascatype"=>"extended", "content"=>"Construction & Building Technology"}, {"ascatype"=>"extended", "content"=>"Science & Technology - Other Topics"}, {"ascatype"=>"extended", "content"=>"Energy & Fuels"}], "count"=>6}, "headings"=>{"heading"=>"Science & Technology", "count"=>1}}, "normalized_languages"=>{"count"=>1, "language"=>{"type"=>"primary", "content"=>"English"}}, "languages"=>{"count"=>1, "language"=>{"type"=>"primary", "content"=>"English"}}, "keywords"=>{"count"=>6, "keyword"=>["Urban water", "Regional water management", "Resilience", "Water supplies", "Water demand", "Supply diversity"]}, "refs"=>{"count"=>73}, "reprint_addresses"=>{"count"=>3, "address_name"=>[{"names"=>{"count"=>1, "name"=>{"seq_no"=>1, "role"=>"author", "full_name"=>"Gonzales, Patricia", "reprint"=>"Y", "addr_no"=>1, "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "wos_standard"=>"Gonzales, P", "first_name"=>"Patricia"}}, "address_spec"=>{"zip"=>{"location"=>"AP", "content"=>94305}, "country"=>"USA", "city"=>"Stanford", "street"=>"473 Via Ortega,Room 314", "addr_no"=>1, "organizations"=>{"organization"=>["Stanford Univ", {"pref"=>"Y", "content"=>"Stanford University"}], "count"=>2}, "full_address"=>"Stanford Univ, Dept Civil & Environm Engn, 473 Via Ortega,Room 314, Stanford, CA 94305 USA", "state"=>"CA", "suborganizations"=>{"count"=>1, "suborganization"=>"Dept Civil & Environm Engn"}}}, {"names"=>{"count"=>1, "name"=>{"seq_no"=>1, "role"=>"author", "full_name"=>"Gonzales, Patricia", "reprint"=>"Y", "addr_no"=>2, "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "wos_standard"=>"Gonzales, P", "first_name"=>"Patricia"}}, "address_spec"=>{"zip"=>{"location"=>"AP", "content"=>94305}, "country"=>"USA", "city"=>"Stanford", "street"=>"473 Via Ortega,Room 117", "addr_no"=>2, "organizations"=>{"organization"=>["Stanford Univ", {"pref"=>"Y", "content"=>"Stanford University"}], "count"=>2}, "full_address"=>"Stanford Univ, ReNUWIt Engn Res Ctr, 473 Via Ortega,Room 117, Stanford, CA 94305 USA", "state"=>"CA", "suborganizations"=>{"count"=>1, "suborganization"=>"ReNUWIt Engn Res Ctr"}}}, {"names"=>{"count"=>1, "name"=>{"seq_no"=>1, "role"=>"author", "full_name"=>"Gonzales, Patricia", "reprint"=>"Y", "addr_no"=>3, "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "wos_standard"=>"Gonzales, P", "first_name"=>"Patricia"}}, "address_spec"=>{"zip"=>{"location"=>"AP", "content"=>94305}, "country"=>"USA", "city"=>"Stanford", "street"=>"473 Via Ortega,Room 204", "addr_no"=>3, "full_address"=>"473 Via Ortega,Room 204, Stanford, CA 94305 USA", "state"=>"CA"}}]}, "abstracts"=>{"count"=>1, "abstract"=>{"abstract_text"=>{"p"=>"The water sector is going through a paradigm shift. Many communities are incorporating decentralized solutions such as water reuse and recycling, stormwater capture, and demand-side management in order to address both short-term and long-term water resources challenges due to population increase, economic growth, intensified climate variability, as well as environmental concerns. For these projects to be sustainable, local characteristics including social and institutional contexts must be incorporated in the planning process. This paper presents a flexible and bottom-up framework that facilitates integration of such characteristics in evaluation of various water resource management strategies. It incorporates various locally-driven factors such as water use efficiency, stress on existing supplies, and adaptation capacity potential, to identify how various local and regional solutions affect resiliency at the utility and regional levels. Rather than defining top-down resilience standards, this framework incorporates quantitative and qualitative assessments that can help decision-makers tailor adaptation measures to the needs and opportunities of a given location or community. A case study application of the framework in the San Francisco Bay Area highlights how community-level characteristics can be used to identify opportunities and adaptation strategies in order to enhance both local and collective water resource resiliency. (C) 2017 Elsevier Ltd. All rights reserved.", "count"=>1}}}, "fund_ack"=>{"grants"=>{"count"=>2, "grant"=>[{"grant_ids"=>{"grant_id"=>"EEC-1028968", "count"=>1}, "grant_agency"=>"National Science Foundation (NSF) Engineering Research Center for Reinventing the Nation's Urban Water Infrastructure (ReNUWIt)"}, {"grant_agency"=>"NSF Graduate Research Fellowship"}]}, "fund_text"=>{"p"=>"This work was supported by the National Science Foundation (NSF) Engineering Research Center for Reinventing the Nation's Urban Water Infrastructure (ReNUWIt) EEC-1028968, and an NSF Graduate Research Fellowship. We would like to thank Adrianne Carr, Michael Hurley, and Andree Johnson for providing us with constructive insights throughout this research."}}, "normalized_doctypes"=>{"doctype"=>"Article", "count"=>1}}, "contributors"=>{"contributor"=>[{"name"=>{"seq_no"=>1, "orcid_id"=>"0000-0002-0956-8001", "role"=>"researcher_id", "full_name"=>"Gonzales, Patricia", "last_name"=>"Gonzales", "display_name"=>"Gonzales, Patricia", "r_id"=>"O-3352-2017", "first_name"=>"Patricia"}}, {"name"=>{"seq_no"=>2, "orcid_id"=>"0000-0003-4421-3764", "role"=>"researcher_id", "full_name"=>"Ajami, newsha", "last_name"=>"Ajami", "display_name"=>"Ajami, newsha", "r_id"=>"C-9151-2017", "first_name"=>"newsha"}}], "count"=>2}}, "r_id_disclaimer"=>"ResearcherID data provided by Clarivate Analytics", "dynamic_data"=>{"citation_related"=>{"tc_list"=>{"silo_tc"=>{"coll_id"=>"WOS", "local_count"=>6}}}, "cluster_related"=>{"identifiers"=>{"identifier"=>[{"type"=>"issn", "value"=>"2210-6707"}, {"type"=>"eissn", "value"=>"2210-6715"}, {"type"=>"doi", "value"=>"10.1016/j.scs.2017.01.012"}]}}}}
    Exception: KeyError: key not found: "organizations"
    lib/rialto/etl/configs/wos_to_sparql_statements.rb:37:in `fetch'

Traceback (most recent call last):
	22: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `block in create_worker'
	21: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:319:in `catch'
	20: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `block (2 levels) in create_worker'
	19: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:320:in `loop'
	18: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:337:in `block (3 levels) in create_worker'
	17: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/concurrent-ruby-1.0.5/lib/concurrent/executor/ruby_thread_pool_executor.rb:348:in `run_task'
	16: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/thread_pool.rb:123:in `block in maybe_in_thread_pool'
	15: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer.rb:554:in `block (3 levels) in process'
	14: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer.rb:429:in `map_to_context!'
	13: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer.rb:429:in `each'
	12: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer.rb:435:in `block in map_to_context!'
	11: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer.rb:476:in `handle_mapping_errors'
	10: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer.rb:436:in `block (2 levels) in map_to_context!'
	 9: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer/step.rb:135:in `execute'
	 8: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer/step.rb:135:in `each'
	 7: from /Users/jlittman/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/traject-3.0.0/lib/traject/indexer/step.rb:138:in `block in execute'
	 6: from lib/rialto/etl/configs/wos_to_sparql_statements.rb:117:in `block (2 levels) in load_config_file'
	 5: from lib/rialto/etl/configs/wos_to_sparql_statements.rb:28:in `fetch_addresses'
	 4: from lib/rialto/etl/configs/wos_to_sparql_statements.rb:28:in `each_with_object'
	 3: from lib/rialto/etl/configs/wos_to_sparql_statements.rb:28:in `each'
	 2: from lib/rialto/etl/configs/wos_to_sparql_statements.rb:30:in `block in fetch_addresses'
	 1: from lib/rialto/etl/configs/wos_to_sparql_statements.rb:37:in `parse_address'
lib/rialto/etl/configs/wos_to_sparql_statements.rb:37:in `fetch': key not found: "organizations" (KeyError)
```

--- a/lib/rialto/etl/configs/wos_to_sparql_statements.rb
+++ b/lib/rialto/etl/configs/wos_to_sparql_statements.rb
@@ -32,18 +32,22 @@ def fetch_addresses(json)
   end
 end
 
+# rubocop:disable Metrics/MethodLength
 def parse_address(addr)
   result = addr.slice('country')
-  org = addr.fetch('organizations').fetch('organization')
-  label = case org
-          when String
-            org
-          when Array
-            org.find { |o| o['pref'] == 'Y' }['content']
-          end
-  result['organization'] = label
+  if addr.key?('organizations')
+    org = addr.fetch('organizations').fetch('organization')
+    label = case org
+            when String
+              org
+            when Array
+              org.find { |o| o['pref'] == 'Y' }['content']
+            end
+    result['organization'] = label
+  end
   result
 end
+# rubocop:enable Metrics/MethodLength
 
 # Find the grants
 def fetch_grant_agencies(json)

--- a/spec/configs/stanford_grants_to_sparql_statements_spec.rb
+++ b/spec/configs/stanford_grants_to_sparql_statements_spec.rb
@@ -96,6 +96,11 @@ RSpec.describe 'stanford_grants_to_sparql_statements' do
   end
 
   before do
+    Settings.entity_resolver.api_key = 'abc123'
+    Settings.entity_resolver.url = 'http://127.0.0.1:3001'
+    # Makes sure Entity Resolver is using above settings.
+    Rialto::Etl::ServiceClient::EntityResolver.instance.initialize_connection
+
     stub_request(:get, 'http://127.0.0.1:3001/organization?name=The%20William%20and%20Flora%20Hewlett%20Foundation')
       .to_return(status: 200,
                  body: organization_uri.to_s,

--- a/spec/configs/wos_to_sparql_statements_spec.rb
+++ b/spec/configs/wos_to_sparql_statements_spec.rb
@@ -540,6 +540,40 @@ RSpec.describe Rialto::Etl::Transformer do
 
       it { is_expected.to eq(1 => { 'country' => 'USA', 'organization' => 'Stanford University' }) }
     end
+    context 'with an address without organizations' do
+      let(:json) do
+        <<~JSON2
+          {
+            "UID": "WOS:000359895400001",
+            "static_data": {
+              "fullrecord_metadata": {
+                "addresses": {
+                  "address_name": [{
+                    "address_spec": {
+                      "zip": {
+          							"location": "AP",
+          							"content": 94305
+          						},
+          						"country": "USA",
+          						"city": "Stanford",
+          						"addr_no": 1,
+          						"full_address": "Stanford Univ, Stanford Cardiovasc Inst, Sch Med, Stanford, CA 94305 USA",
+          						"state": "CA",
+          						"suborganizations": {
+          							"count": 2,
+          							"suborganization": ["Stanford Cardiovasc Inst", "Sch Med"]
+          						}
+                    }
+                  }]
+                }
+              }
+            }
+          }
+        JSON2
+      end
+
+      it { is_expected.to eq(1 => { 'country' => 'USA' }) }
+    end
   end
   describe '#fetch_grant_agencies' do
     subject { indexer.fetch_grant_agencies(json) }

